gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientGrpcDownscopingInterceptor.java

```diff
@@ -0,0 +1,224 @@
+    this.downscopingFunction = downscopedAccessToken;
+  }
+
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+      MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+    logger.atFinest().log(String.format("interceptCall(): method=%s", method.getFullMethodName()));
+    return new SimpleForwardingClientCall<ReqT, RespT>(next.newCall(method, callOptions)) {
+      private int flowControlRequests;
+      private final String methodName = method.getFullMethodName();;
+      private Metadata headers;
+      private Listener<RespT> responseListener;
+
+      @Override
+      public void start(Listener<RespT> responseListener, Metadata headers) {
+        this.responseListener = responseListener;
+        this.headers = headers;
+
+        logger.atFinest().log("start(): method=%s", methodName);
+      }
+
+      @Override
+      public void request(int numMessages) {
+        if (headers != null) {
+          this.flowControlRequests += numMessages;
+        } else {
+          super.request(numMessages);
+        }
+      }
+
+      @Override
+      public void sendMessage(ReqT message) {
+        if (headers != null) {
+          setAuthHeader(message);
+          super.start(responseListener, headers);
+          headers = null;
+          if (flowControlRequests != 0) {
+            super.request(flowControlRequests);
+            flowControlRequests = 0;
+          }
+        }
+        super.sendMessage(message);
+      }
+
+      private void setAuthHeader(ReqT message) {
+        if (headers == null) {
+          return;
+        }
+
+        String token = getDownScopedToken(message);
+        if (token != null) {
+          logger.atFinest().log("Setting down-scoped auth token");
+          headers.put(AUTH_KEY, "Bearer " + token);
+        }
+      }
+
+      private String getDownScopedToken(ReqT message) {
+        List<AccessBoundary> accessBoundaries = getAccessBoundaries(message);
+        return getDownscopedToken(accessBoundaries);
+      }
+
+      private String getBucketName(String bucketLongName) {
+        return bucketLongName.substring(BUCKET_PREFIX_LENGTH);
+      }
+
+      private List<AccessBoundary> getAccessBoundaries(ReqT message) {
+        if (GOOGLE_STORAGE_V_2_STORAGE_READ_OBJECT.equals(methodName)) {
+          ReadObjectRequest readObjectRequest = (ReadObjectRequest) message;
+          String bucketName = getBucketName(readObjectRequest.getBucket());
+          String objectName = readObjectRequest.getObject();
+
+          return ImmutableList.of(
+              AccessBoundary.create(bucketName, objectName, Action.READ_OBJECTS));
+        }
+
+        if (GOOGLE_STORAGE_V_2_STORAGE_WRITE_OBJECT.equals(methodName)) {
+          WriteObjectSpec writeObjectSpec = ((WriteObjectRequest) message).getWriteObjectSpec();
+          com.google.storage.v2.Object obj = writeObjectSpec.getResource();
```

Comment:
@frankyn can you confirm whether we will have Resource attached in WriteObjectSpec for ResumableUpload specific writeObject. By looking into the code and verification via integration test I am able to reach the conclusion that,
*  for Direct Uplaod request (i.e. part file up-load in PCU) WriteObjectRequest have `FirstMessageCase` as 2 (i.e. WRITE_OBJECT_SPEC type) and that contains Resource info
* for resumable upload WriteObjectRequest have `FirstMessageCase` as 1 (i.e. UPLOAD_ID) and do not contain Resource info.

IIUC, object and Bucket extracted here would have empty string value (`""`).

Reply:
@singhravidutt, I think you're referring to the following code block:

```
if (flowControlRequests != 0) {
            super.request(flowControlRequests);
            flowControlRequests = 0;
}
```

This flowControl code was provided by @ejona86; Eric, can you clarify how this variable is used to make sure headers in the initial message are modified? IIUC only the initial message requires the Auth header and not subsequent messages.

Reply:
After our chat, my current hypothesis as to why interceptor works without bucket or object names is that when `upload_id` is provided in resumable upload request in gRPC, auth header is not required similar to [JSON/XML APIs](https://cloud.google.com/storage/docs/performing-resumable-uploads#initiate-session).

That said, we need to confirm with @brandony, does gRPC work similar to JSON/XML that uploading to a resumable session only requires upload_id once the session is created?

![image](https://github.com/GoogleCloudDataproc/hadoop-connectors/assets/1907648/9ea8c15e-85dc-44f2-9869-a260c01f167d)

Reply:
given that @arunkumarchacko is able to test it out even for resumable upload. Going ahead with the current changes.

Reply:
You're correct. The upload ID is the credential; no need for an auth header.